sim/dex-data.ts

```diff
@@ -34,7 +34,21 @@ export function toID(text: any): ID {
 	/* eslint-enable @typescript-eslint/prefer-optional-chain */
 }
 
-export class BasicEffect implements EffectData {
+/**
+ * Like Object.assign but only assigns fields missing from self.
+ * Facilitates consistent field ordering in constructors.
+ * Modifies self in-place.
+ */
+export function assignNewFields(self: AnyObject, data: AnyObject) {
+	for (const k in data) {
+		// to behave like Object.assign, skip data's inherited properties
+		if (!Object.prototype.hasOwnProperty.call(data, k)) continue;
+		if (Object.prototype.hasOwnProperty.call(self, k)) continue;
```

Comment:
Unless there are performance implications, this would be more readable as `k in self` - we shouldn't be doing weird things to the prototype chain.